File: CS50x-pset-1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    //initialize values
    long number = 0;
    int i = 0;
    int length = 0;
    int add = 0;
    int total = 0;
    bool checksum = false;

    //prompt user for credit card number
    do
    {
        number = get_long("Number: ");
    }
    while (number < 0); //reprompt if negative number is inputted

    //calculate length
    while (i <= length)
    {
        if (number > (pow(10, i)))
        {
            length++;
        }
        i++;
    }

    //calculate checksum
    //multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits together
    for (i = 2; i <= length; i += 2)
    {
        add = (number % (long)(pow(10, i))) / (pow(10, i - 1));
        if ((add * 2 / 10) == 0)
        {
            total += add * 2;
        }
        else
        {
            total += (((add * 2) % 10) + 1);
        }
    }

    //add the sum to the sum of the digits that weren’t multiplied by 2
    for (i = 1; i <= length; i += 2)
    {
        add = (number % (long)(pow(10, i))) / (pow(10, i - 1));
        total += add;
    }

    //check if checksum is valid
    if ((total % 10) == 0)
    {
        checksum = true;
    }

    //check for card length and starting digits then print result
    if (checksum)
    {
        if (length == 15 && (int)((number % (long)(pow(10, 15))) / (pow(10, 14))) == 3
            && ((int)((number % (long)(pow(10, 14))) / (pow(10, 13))) == 4 || (int)((number % (long)(pow(10, 14))) / (pow(10, 13))) == 7))
        {
            printf("AMEX\n");
        }
        else if (length == 16 && (int)((number % (long)(pow(10, 16))) / (pow(10, 15))) == 5
                 && ((int)((number % (long)(pow(10, 15))) / (pow(10, 14))) == 1 || (int)((number % (long)(pow(10, 15))) / (pow(10, 14))) == 2
                     || (int)((number % (long)(pow(10, 15))) / (pow(10, 14))) == 3 || (int)((number % (long)(pow(10, 15))) / (pow(10, 14))) == 4
                     || (int)((number % (long)(pow(10, 15))) / (pow(10, 14))) == 5))
        {
            printf("MASTERCARD\n");
        }
        else if (length == 13 && (int)((number % (long)(pow(10, 13))) / (pow(10, 12))) == 4)
        {
            printf("VISA\n");
        }
        else if (length == 16 && (int)((number % (long)(pow(10, 16))) / (pow(10, 15))) == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: CS50x-pset-6/credit/credit.py
from cs50 import get_int
import math

# prompt user for credit card number until positive number is inputted
while(True):
    number = get_int("Number: ")
    if (number > 0):
        break

# calculate length
i = 0
length = 0
while(i <= length):
    if number > pow(10, i):
        length += 1
    i += 1

# calculate checksum
# multiply every other digit by 2, starting with the number's second-to-last digit, and then add those products' digits together
total = 0
for i in range(2, length + 2, 2):
    add = ((number % pow(10, i)) / pow(10, i - 1))
    add = int(add)
    if math.floor(add * 2 / 10) == 0:
        total += (add * 2)
    else:
        total += (((add * 2) % 10) + 1)

# add the sum to the sum of the digits that weren't multiplied by 2
for i in range(1, length + 2, 2):
    add = ((number % pow(10, i)) / pow(10, i - 1))
    total += add
    total = int(total)

# check if checksum is valid
checksum = False
if total % 10 == 0:
    checksum = True

# check for card length and starting digits then print result
if checksum:
    if length == 15 and math.floor((number % pow(10, 15)) / pow(10, 14)) == 3 and math.floor(((number % pow(10, 14)) / pow(10, 13)) == 4 or math.floor((number % pow(10, 14)) / pow(10, 13)) == 7):
        print("AMEX")
    elif length == 16 and math.floor((number % pow(10, 16)) / pow(10, 15)) == 5 and math.floor((number % pow(10, 15)) / pow(10, 14)) <= 5 and math.floor((number % pow(10, 15)) / pow(10, 14)) >= 1:
        print("MASTERCARD")
    elif length == 13 and math.floor((number % pow(10, 13)) / pow(10, 12)) == 4:
        print("VISA")
    elif length == 16 and math.floor((number % pow(10, 16)) / pow(10, 15)) == 4:
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")